paddle/fluid/operators/mkldnn/mul_mkldnn_op.cc

```diff
@@ -0,0 +1,518 @@
+            memory({dst_mdesc, engine_}, to_void_cast<K>(x_tmp.data<K>()));
+
+        auto reorder = mkldnn::reorder(src_mem, dst_mem);
+        stream(stream::kind::eager).submit({reorder}).wait();
+      } else {
+        x_tmp.mutable_data<T>(ctx.GetPlace(), data->memory_size());
+
+        auto src_mem =
+            memory({src_mdesc, engine_}, to_void_cast<T>(data->data<T>()));
+        auto dst_mem =
+            memory({dst_mdesc, engine_}, to_void_cast<T>(x_tmp.data<T>()));
+
+        auto reorder = mkldnn::reorder(src_mem, dst_mem);
+        stream(stream::kind::eager).submit({reorder}).wait();
+      }
+
+      x_tmp.Resize(data->dims());
+      x_tmp.set_format((memory::format)dst_mdesc.data.format);
+      data_matrix = framework::ReshapeToMatrix(x_tmp, num_col_dims);
+    } else {
+      data_matrix = framework::ReshapeToMatrix(*data, num_col_dims);
+    }
+
+    return data_matrix;
+  }
+
+  memory Reorder(const memory::desc &src_desc, const memory::desc &dst_desc,
+                 const void *src_data) {
+    auto src_mem = memory({src_desc, engine_}, const_cast<void *>(src_data));
+    auto dst_mem = memory({dst_desc, engine_});
+
+    auto reorder = mkldnn::reorder(src_mem, dst_mem);
+    stream(stream::kind::eager).submit({reorder}).wait();
+
+    return dst_mem;
+  }
+
+  memory TransposeInputY(const Tensor *input_y) {
+    auto dims = framework::vectorize2int(input_y->dims());
+    std::swap(dims[0], dims[1]);  // Correct output dimensions
+    auto src_desc = CreateMemDescriptor<K>(dims, memory::format::io);
+    auto dst_desc = CreateMemDescriptor<K>(dims, memory::format::oi);
+    return Reorder(src_desc, dst_desc, input_y->data<K>());
+  }
+
+  template <typename DT>
+  memory::desc CreateMemDescriptor(const std::vector<int> &dims,
+                                   memory::format format) {
+    return platform::MKLDNNMemDesc(dims, platform::MKLDNNGetDataType<DT>(),
+                                   format);
+  }
+
+  template <typename DT>
+  memory::desc CreateMemDescriptor(const Tensor *tensor,
+                                   memory::format format) {
+    auto dims = framework::vectorize2int(tensor->dims());
+    return CreateMemDescriptor<DT>(dims, format);
+  }
+
+  memory::desc CreateMemDescriptor(const std::vector<int> &dims,
+                                   memory::data_type type,
+                                   memory::format format) {
+    return platform::MKLDNNMemDesc(dims, type, format);
+  }
+
+  memory::desc CreateMemDescriptor(const Tensor *tensor, memory::data_type type,
+                                   memory::format format) {
+    auto dims = framework::vectorize2int(tensor->dims());
+    return CreateMemDescriptor(dims, type, format);
+  }
+
+  memory CreateMemory(const memory::desc &desc, const Tensor *tensor,
+                      bool is_y) {
+    if (is_y)
+      return CreateMemory(desc, tensor->data<K>());
+    else
+      return CreateMemory(desc, tensor->data<T>());
+  }
```

Comment:
The `is_y` parameter is unnecessary here. It'd be enough to templatize the function with datatype that would be passed to tensor->data<T>() function.

```
  template <typename DT>
  memory CreateMemory(const memory::desc &desc, const Tensor *tensor) {
      return CreateMemory(desc, tensor->data<DT>());
  }
```